contracts/test/Nodes.sol

```diff
@@ -0,0 +1,159 @@
+        return
+            Nodes.Node({
+                signingKeyPub: _genBytes(32),
+                httpAddress: _genString(32),
+                isHealthy: isHealthy
+            });
+    }
+
+    function test_canAddNode() public {
+        Nodes.Node memory node = _randomNode(true);
+
+        address operatorAddress = vm.randomAddress();
+
+        uint16 nodeId = nodes.addNode(
+            operatorAddress,
+            node.signingKeyPub,
+            node.httpAddress
+        );
+
+        vm.assertEq(nodes.ownerOf(nodeId), operatorAddress);
+        vm.assertEq(nodes.getNode(nodeId).signingKeyPub, node.signingKeyPub);
+        vm.assertEq(nodes.getNode(nodeId).httpAddress, node.httpAddress);
+        vm.assertEq(nodes.getNode(nodeId).isHealthy, true);
+    }
+
+    function test_canAddMultiple() public {
+        Nodes.Node memory node1 = _randomNode(true);
+        Nodes.Node memory node2 = _randomNode(true);
+        Nodes.Node memory node3 = _randomNode(true);
+
+        address operator1 = vm.randomAddress();
+        address operator2 = vm.randomAddress();
+        address operator3 = vm.randomAddress();
+
+        uint16 node1Id = nodes.addNode(
+            operator1,
+            node1.signingKeyPub,
+            node1.httpAddress
+        );
+        nodes.addNode(operator2, node2.signingKeyPub, node2.httpAddress);
+        nodes.addNode(operator3, node3.signingKeyPub, node3.httpAddress);
+
+        Nodes.NodeWithId[] memory healthyNodes = nodes.healthyNodes();
+        vm.assertTrue(healthyNodes.length == 3);
+
+        nodes.updateHealth(node1Id, false);
+        healthyNodes = nodes.healthyNodes();
+        vm.assertTrue(healthyNodes.length == 2);
+    }
+
+    function test_canMarkUnhealthy() public {
+        Nodes.Node memory node = _randomNode(true);
+        address operator = vm.randomAddress();
+
+        uint16 nodeId = nodes.addNode(
+            operator,
+            node.signingKeyPub,
+            node.httpAddress
+        );
+
+        nodes.updateHealth(nodeId, false);
+
+        vm.assertEq(nodes.getNode(nodeId).isHealthy, false);
+        vm.assertEq(nodes.healthyNodes().length, 0);
+    }
+
+    function testFail_ownerCannotUpdateHealth() public {
+        vm.expectRevert(Ownable.OwnableUnauthorizedAccount.selector);
+        Nodes.Node memory node = _randomNode(true);
+        address operator = vm.randomAddress();
+
+        uint16 nodeId = nodes.addNode(
+            operator,
+            node.signingKeyPub,
+            node.httpAddress
+        );
+
+        vm.prank(operator);
+        nodes.updateHealth(nodeId, false);
+    }
+
+    function testFail_ownerCannotTransfer() public {
+        Nodes.Node memory node = _randomNode(true);
+        address operator = vm.randomAddress();
+
+        uint16 nodeId = nodes.addNode(
+            operator,
+            node.signingKeyPub,
+            node.httpAddress
+        );
+
+        vm.prank(operator);
```

Comment:
`vm.prank` is cool. It just lets you change the `msg.sender` value to any other address for a test, so you can impersonate someone and see how the contract behaves.